Rework error message collection in structured Dict

lafrech opened the pull request:
I believe this fixes #730.

@shabble, @deckar01, would you mind reviewing?

Thanks.

lafrech committed:
Rework error message collection in structured Dict

Fixes #730

deckar01 commented on marshmallow/fields.py:
```diff
@@ -1195,31 +1195,23 @@ def _deserialize(self, value, attr, data):
-                        'Invalid key: {}'.format(message)
-                        for message in e.messages
-                    ]
+                    errors[key]['key'] = e.messages
```
~~I don't think changing the error structure is necessary to fix this bug. From what I can tell, the dict deserialization error structure I implemented is closer to the list deserialization structure than what was proposed in https://github.com/marshmallow-code/marshmallow/issues/730#issuecomment-361362161.~~

I actually like how using a dictionary simplifies the implementation by avoiding array extension and text prefixes, ~~but I propose moving changes to the error structure to a separate issue/PR~~.

deckar01 replied:
I see why it is needed now. Thanks for digging into this @lafrech.

deckar01 replied:
It might be helpful to provide a more technical insight about this fix in the PR description for subsequent reviewers. It wasn't clear from the conversation in the issue why this was a necessary.

deckar01 commented on marshmallow/fields.py:
```diff
@@ -1195,31 +1195,23 @@ def _deserialize(self, value, attr, data):
-                        errors[key] = []
-                    errors[key].extend([
-                        'Invalid value: {}'.format(message)
-                        for message in e.messages
```
Er, wait. The `messages` property of a `ValidationError` can be a "string, list, or dictionary of error messages".

https://marshmallow.readthedocs.io/en/latest/api_reference.html#marshmallow.exceptions.ValidationError.messages

Ya, I was assuming it was a list. We can't safely merge the value and key errors, so making it a dict to separate them makes sense.

deckar01 commented on tests/test_deserialization.py:
```diff
@@ -1304,6 +1303,60 @@ class MySchema(Schema):
+                    'birthdate': ['Not a valid datetime.']}},
+                ('whatever', ): {
+                    'key': ['Not a valid string.'],
+                    'value': {'birthdate': ['Not a valid datetime.']}}
```
This test seems really big. Is it possible to boil the repro case down further? Any `ValidationError` that returns a dict for `messages` should cover the bug.

lafrech replied:
You're right. This test is more a demo than a test. I just shortened it. I kept a success case and a failure. Do you think only the failure is relevant to the test?

deckar01 replied:
It looks like we only need to test the error from of custom validators. I was referring to the size of the data structure not the number of cases. Sorry for not being clear.

If this can be reproduced without using a nested schema it would be easier to follow. A schema with a single field that has a custom validator that either produces an error string or error dict should cover the bug.

I am OK with the tests as they are now since they match the reported issue. Any additional work would be icing on the cake. Thanks for working on this.

lafrech committed:
Rework test_structured_dict_nested_value_deserialization

lafrech committed:
Rework test_structured_dict_nested_value_deserialization

lafrech committed:
Structured Dict: test field returning dict of error messages

Reg test for #730

lafrech committed:
Remove test_structured_dict_nested_value_deserialization

This is now covered by test_structured_dict_value_deserialization and
test_structured_dict_key_deserialization

lafrech commented:
Thanks @deckar01 for the review. I think I managed to get it right and minimal.

deckar01 commented:
Looks good to me. :+1:

lafrech commented:
@sloria, I think we're done with this.

sloria commented:
Awesome!
